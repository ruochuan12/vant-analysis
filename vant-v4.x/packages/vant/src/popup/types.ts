import type { Ref, ComponentPublicInstance } from 'vue';
import type { PopupProps } from './Popup';

export type PopupPosition = 'top' | 'left' | 'bottom' | 'right' | 'center' | '';

export type PopupCloseIconPosition =
  | 'top-left'
  | 'top-right'
  | 'bottom-left'
  | 'bottom-right';

export type PopupExpose = {
  popupRef: Ref<HTMLElement>;
};

export type PopupInstance = ComponentPublicInstance<PopupProps, PopupExpose>;

export type PopupThemeVars = {
  popupBackground?: string;
  popupTransition?: string;
  popupRoundRadius?: string;
  popupCloseIconSize?: string;
  popupCloseIconColor?: string;
  popupCloseIconMargin?: string;
  popupCloseIconZIndex?: number | string;
};
