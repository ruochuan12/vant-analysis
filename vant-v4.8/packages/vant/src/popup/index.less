:root,
:host {
  --van-popup-background: var(--van-background-2);
  --van-popup-transition: transform var(--van-duration-base);
  --van-popup-round-radius: 16px;
  --van-popup-close-icon-size: 22px;
  --van-popup-close-icon-color: var(--van-gray-5);
  --van-popup-close-icon-margin: 16px;
  --van-popup-close-icon-z-index: 1;
}

.van {
  &-overflow-hidden {
    overflow: hidden !important;
  }

  &-popup {
    position: fixed;
    max-height: 100%;
    overflow-y: auto;
    box-sizing: border-box;
    background: var(--van-popup-background);
    transition: var(--van-popup-transition);
    -webkit-overflow-scrolling: touch;

    &--center {
      top: 50%;
      left: 0;
      right: 0;
      width: fit-content;
      max-width: calc(100vw - var(--van-padding-md) * 2);
      margin: 0 auto;
      transform: translateY(-50%);

      &.van-popup--round {
        border-radius: var(--van-popup-round-radius);
      }
    }

    &--top {
      top: 0;
      left: 0;
      width: 100%;

      &.van-popup--round {
        border-radius: 0 0 var(--van-popup-round-radius)
          var(--van-popup-round-radius);
      }
    }

    &--right {
      top: 50%;
      right: 0;
      transform: translate3d(0, -50%, 0);

      &.van-popup--round {
        border-radius: var(--van-popup-round-radius) 0 0
          var(--van-popup-round-radius);
      }
    }

    &--bottom {
      bottom: 0;
      left: 0;
      width: 100%;

      &.van-popup--round {
        border-radius: var(--van-popup-round-radius)
          var(--van-popup-round-radius) 0 0;
      }
    }

    &--left {
      top: 50%;
      left: 0;
      transform: translate3d(0, -50%, 0);

      &.van-popup--round {
        border-radius: 0 var(--van-popup-round-radius)
          var(--van-popup-round-radius) 0;
      }
    }

    &-slide-top-enter-active,
    &-slide-left-enter-active,
    &-slide-right-enter-active,
    &-slide-bottom-enter-active {
      transition-timing-function: var(--van-ease-out);
    }

    &-slide-top-leave-active,
    &-slide-left-leave-active,
    &-slide-right-leave-active,
    &-slide-bottom-leave-active {
      transition-timing-function: var(--van-ease-in);
    }

    &-slide-top-enter-from,
    &-slide-top-leave-active {
      transform: translate3d(0, -100%, 0);
    }

    &-slide-right-enter-from,
    &-slide-right-leave-active {
      transform: translate3d(100%, -50%, 0);
    }

    &-slide-bottom-enter-from,
    &-slide-bottom-leave-active {
      transform: translate3d(0, 100%, 0);
    }

    &-slide-left-enter-from,
    &-slide-left-leave-active {
      transform: translate3d(-100%, -50%, 0);
    }

    &__close-icon {
      position: absolute;
      z-index: var(--van-popup-close-icon-z-index);
      color: var(--van-popup-close-icon-color);
      font-size: var(--van-popup-close-icon-size);

      &--top-left {
        top: var(--van-popup-close-icon-margin);
        left: var(--van-popup-close-icon-margin);
      }

      &--top-right {
        top: var(--van-popup-close-icon-margin);
        right: var(--van-popup-close-icon-margin);
      }

      &--bottom-left {
        bottom: var(--van-popup-close-icon-margin);
        left: var(--van-popup-close-icon-margin);
      }

      &--bottom-right {
        right: var(--van-popup-close-icon-margin);
        bottom: var(--van-popup-close-icon-margin);
      }
    }
  }
}
