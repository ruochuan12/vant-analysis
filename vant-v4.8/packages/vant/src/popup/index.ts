import { withInstall } from '../utils';
import _Popup from './Popup';

export const Popup = withInstall(_Popup);
export default Popup;
export { popupProps } from './Popup';
export type { PopupProps } from './Popup';
export type {
  PopupPosition,
  PopupInstance,
  PopupThemeVars,
  PopupCloseIconPosition,
} from './types';

declare module 'vue' {
  export interface GlobalComponents {
    VanPopup: typeof Popup;
  }
}
