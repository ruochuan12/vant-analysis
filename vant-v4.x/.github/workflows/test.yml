name: CI

on:
  push:
    branches:
      - '**'

  pull_request:
    branches:
      - main

  workflow_dispatch:

jobs:
  lint:
    runs-on: ubuntu-latest
    steps:
    - uses: actions/checkout@v4

    - name: Install pnpm
      run: corepack enable

    - uses: actions/setup-node@v3
      with:
        node-version: '18'
        cache: 'pnpm'

    - name: Install dependencies
      run: pnpm install

    - name: Run linter
      run: npm run lint

  test:
    runs-on: macos-14
    steps:
    - uses: actions/checkout@v4

    - name: Install pnpm
      run: corepack enable

    - uses: actions/setup-node@v3
      with:
        node-version: '18'
        cache: 'pnpm'

    - name: Install dependencies
      run: pnpm install

    - name: Run test cases
      run: pnpm run test:coverage

    - name: Upload coverage to Codecov  
      uses: codecov/codecov-action@v3
      with:
        token: ${{ secrets.CODECOV_TOKEN }}
        directory: ./packages/vant/test/coverage

  build:
    runs-on: macos-14
    steps:
    - uses: actions/checkout@v4

    - name: Install pnpm
      run: corepack enable

    - uses: actions/setup-node@v3
      with:
        node-version: '18'
        cache: 'pnpm'

    - name: Install dependencies
      run: pnpm install

    - name: Build
      run: npm run build
