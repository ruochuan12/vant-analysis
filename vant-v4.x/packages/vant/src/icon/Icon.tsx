import {
  inject,
  computed,
  defineComponent,
  type PropType,
  type ExtractPropTypes,
} from 'vue';
import {
  addUnit,
  numericProp,
  makeStringProp,
  createNamespace,
} from '../utils';
import { Badge, type BadgeProps } from '../badge';
import { CONFIG_PROVIDER_KEY } from '../config-provider/ConfigProvider';

const [name, bem] = createNamespace('icon');

const isImage = (name?: string) => name?.includes('/');

export const iconProps = {
  dot: Boolean,
  tag: makeStringProp<keyof HTMLElementTagNameMap>('i'),
  name: String,
  size: numericProp,
  badge: numericProp,
  color: String,
  badgeProps: Object as PropType<Partial<BadgeProps>>,
  classPrefix: String,
};

export type IconProps = ExtractPropTypes<typeof iconProps>;

export default defineComponent({
  name,

  props: iconProps,

  setup(props, { slots }) {
    const config = inject(CONFIG_PROVIDER_KEY, null);

    const classPrefix = computed(
      () => props.classPrefix || config?.iconPrefix || bem(),
    );

    return () => {
      const { tag, dot, name, size, badge, color } = props;
      const isImageIcon = isImage(name);

      return (
        <Badge
          dot={dot}
          tag={tag}
          class={[
            classPrefix.value,
            isImageIcon ? '' : `${classPrefix.value}-${name}`,
          ]}
          style={{
            color,
            fontSize: addUnit(size),
          }}
          content={badge}
          {...props.badgeProps}
        >
          {slots.default?.()}
          {isImageIcon && <img class={bem('image')} src={name} />}
        </Badge>
      );
    };
  },
});
