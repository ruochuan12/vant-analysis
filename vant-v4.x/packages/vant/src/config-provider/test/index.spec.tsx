import { ref } from 'vue';
import { ConfigProvider, type ConfigProviderThemeVars } from '..';
import { Icon } from '../../icon';
import { Popup } from '../../popup';
import { later, mount } from '../../../test';

test('should render tag prop correctly', () => {
  const wrapper = mount(ConfigProvider, {
    props: {
      tag: 'section',
    },
  });
  expect(wrapper.html()).toMatchSnapshot();
});

test('should change icon class-prefix when using icon-prefix prop', () => {
  const wrapper = mount({
    render() {
      return (
        <ConfigProvider iconPrefix="foo">
          <Icon name="success" />
        </ConfigProvider>
      );
    },
  });
  expect(wrapper.html()).toMatchSnapshot();
});

test('should change global z-index when using z-index prop', async () => {
  const show = ref(true);
  const wrapper = mount({
    render() {
      return (
        <ConfigProvider zIndex={0}>
          <Popup v-model:show={show.value} />
        </ConfigProvider>
      );
    },
  });

  await later();
  expect(wrapper.find('.van-popup').style.zIndex).toEqual('1');
});

test('should apply theme-vars-light in light mode', () => {
  const wrapper = mount({
    render() {
      return (
        <ConfigProvider
          themeVars={{ rateIconFullColor: 'red' }}
          themeVarsDark={{ rateIconFullColor: 'green' }}
          themeVarsLight={{ rateIconFullColor: 'blue' }}
        />
      );
    },
  });

  expect(wrapper.element.getAttribute('style')).toEqual(
    '--van-rate-icon-full-color: blue;',
  );
});

test('should apply basic theme vars correctly', () => {
  const wrapper = mount({
    render() {
      const themeVars: ConfigProviderThemeVars = {
        gray1: '#111',
        background2: 'red',
      };
      return <ConfigProvider themeVars={themeVars} />;
    },
  });

  expect(wrapper.element.getAttribute('style')).toEqual(
    '--van-gray-1: #111; --van-background-2: red;',
  );
});

test('should apply theme-vars-dark in dark mode', () => {
  const wrapper = mount({
    render() {
      return (
        <ConfigProvider
          theme="dark"
          themeVars={{ rateIconFullColor: 'red' }}
          themeVarsDark={{ rateIconFullColor: 'green' }}
          themeVarsLight={{ rateIconFullColor: 'blue' }}
        />
      );
    },
  });

  expect(wrapper.element.getAttribute('style')).toEqual(
    '--van-rate-icon-full-color: green;',
  );
});

test('should apply theme-vars-scope enable root affects', async () => {
  const wrapper = mount({
    render() {
      return (
        <ConfigProvider
          theme="dark"
          themeVars={{ rateIconFullColor: 'red' }}
          themeVarsDark={{ rateIconFullColor: 'green' }}
          themeVarsLight={{ rateIconFullColor: 'blue' }}
          themeVarsScope="global"
        />
      );
    },
  });

  expect(document.documentElement.getAttribute('style')).toEqual(
    '--van-rate-icon-full-color: green;',
  );
  expect(
    wrapper.element.getAttribute('style') ===
      '--van-rate-icon-full-color: green;',
  ).toBeFalsy();

  await wrapper.setProps({
    themeVarsScope: 'local',
  });

  expect(wrapper.element.getAttribute('style')).toEqual(
    '--van-rate-icon-full-color: green;',
  );
  expect(
    document.documentElement.getAttribute('style') ===
      '--van-rate-icon-full-color: green;',
  ).toBeFalsy();
});

test('should apply theme-vars-scope enable root affects and sync theme vars', async () => {
  const wrapper = mount({
    render() {
      return (
        <ConfigProvider
          theme="dark"
          themeVars={{ rateIconFullColor: 'red' }}
          themeVarsScope="global"
        />
      );
    },
  });

  expect(document.documentElement.getAttribute('style')).toEqual(
    '--van-rate-icon-full-color: red;',
  );

  await wrapper.setProps({
    themeVars: {
      rateIconFullColor: 'red',
      buttonPrimaryColor: 'red',
    },
  });

  expect(document.documentElement.getAttribute('style')).toEqual(
    '--van-rate-icon-full-color: red; --van-button-primary-color: red;',
  );

  await wrapper.setProps({
    themeVars: {
      buttonPrimaryColor: 'red',
    },
  });

  expect(document.documentElement.getAttribute('style')).toEqual(
    '--van-button-primary-color: red;',
  );
});
