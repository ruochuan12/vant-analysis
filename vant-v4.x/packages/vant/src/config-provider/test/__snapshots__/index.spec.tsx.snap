// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`should change icon class-prefix when using icon-prefix prop 1`] = `
<div class="van-config-provider">
  <i class="van-badge__wrapper foo foo-success">
  </i>
</div>
`;

exports[`should render tag prop correctly 1`] = `
"<section class="van-config-provider">
  <!---->
</section>"
`;
